docs/rules/jsx-no-leaked-zero.md

```diff
@@ -0,0 +1,159 @@
+
+This can be avoided by:
+- casting the condition to bool
+- transforming the binary expression into a ternary expression which returns `null` for falsy values
+
+This rule is autofixable, check the Options section to read more about the different strategies available.
+
+Examples of **incorrect** code for this rule:
+
+```jsx
+const Component = ({ count, title }) => {
+  return <div>{count && title}</div>
+}
+```
+
+```jsx
+const Component = ({ count }) => {
+  return <div>{count && <span>There are {count} results</span>}</div>
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{elements.length && <List elements={elements}/>}</div>
+}
+```
+
+```jsx
+const Component = ({ nestedCollection }) => {
+  return (
+    <div>
+      {nestedCollection.elements.length && <List elements={nestedCollection.elements} />}
+    </div>
+  )
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{elements[0] && <List elements={elements}/>}</div>
+}
+```
+
+```jsx
+const Component = ({ numberA, numberB }) => {
+  return <div>{(numberA || numberB) && <Results>{numberA+numberB}</Results>}</div>
+}
+```
+
+```jsx
+// If the condition is a boolean value, this rule will report the logical expression
+// since it can't guess the type of the condition.
+const Component = ({ someBool }) => {
+  return <div>{someBool && <Results>{numberA+numberB}</Results>}</div>
+}
+```
+
+Examples of **correct** code for this rule:
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{elements}</div>
+}
+```
+
+```jsx
+// An OR condition it's considered valid since it's assumed as a way
+// to render some fallback if the first value is falsy, not to render something conditionally.
+const Component = ({ customTitle }) => {
+  return <div>{customTitle || defaultTitle}</div>
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>There are {elements.length} elements</div>
+}
+```
+
+```jsx
+const Component = ({ elements, count }) => {
+  return <div>{!count && 'No results found'}</div>
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{!!elements.length && <List elements={elements}/>}</div>
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{Boolean(elements.length) && <List elements={elements}/>}</div>
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{elements.length > 0 && <List elements={elements}/>}</div>
+}
+```
+
+```jsx
+const Component = ({ elements }) => {
+  return <div>{elements.length ? <List elements={elements}/> : null}</div>
+}
+```
+
+### Options
+
+The supported options are:
+
+### `fixStrategy`
+`"cast"` or `"ternary"` (default: `"ternary"`) - Decide the strategy to autofix the rule. The "cast" option will cast to boolean the condition of the JSX expression. The "ternary" option transforms the binary expression into a ternary expression returning `null` for falsy values.
```

Comment:
this does answer the question, but typically rules don't have autofix-specific options - in other words, i'd expect someone who sets this rule to "cast" would see warnings with a ternary, and vice versa.

Reply:
Not sure if I understood this correctly. Do you mean that code not leaking zeros but using the opposite strategy from the one set in the rule option should be reported as incorrect too?

Reply:
No, what i mean is, a rule should behave identically with or without autofix (except for whether the violations are fixed or not). So, I would expect the rule itself to be configurable to allow a cast, or a ternary, or both. An unleaked zero that's using a disallowed mechanism would warn.

Reply:
> What i mean is, a rule should behave identically with or without autofix (except for whether the violations are fixed or not)

Isn't that the opposite of what you are suggesting then? The current implementation will report the same errors no matter which `fixStrategy` the user picks. The option is just to decide which strategy the user prefers for autofixing the errors.

If the user sets `fixStrategy: "cast"` and there is a ternary inside a JSX expression, it won't be reported, since the possibility of getting a leaked zero is prevented.

Changing the behavior to report unleaked zero using the other option not set in `fixStrategy` would make the rule not to behave identically in every case.

Reply:
If they set their strategy to "cast", then I would expect the rule to warn on a ternary (that, if done as a &&, would leak a zero).

The rule shouldn't behave identically with different options - if the options are different, and the rule behaves differently, then why are there options at all

Reply:
Additionally, what would be the autofix strategy if a third "both" value is added for the `fixStrategy` option?

IMO I think we are overcomplicating this. The goal of this option was to offer 2 different ways of autofixing the same error depending on the user's taste. If that's an issue, I'd just stick to one way of always autofixing the error, and remove the rule option and the other autofix implementation.

Reply:
> The rule shouldn't behave identically with different options - if the options are different, and the rule behaves differently, then why are there options at all

The autofix strategy is behaving differently, that was the purpose of the option. So I'm thinking of just removing it and stick to a ternary autofix.

Reply:
eslint rules indeed don't offer fix-specific options.

I also don't think there's enough value for the rule at all if it's going to only force one of the fix strategies.

Reply:
I really don't know what you are suggesting then... If we don't want to offer an autofix option (I don't know why this is a problem even if other rules don't offer it, it shouldn't prevent this one from offering such an option), but enforcing only one fix strategy makes the rule not have enough value, how am I supposed to solve this?

Reply:
My hope is that you'd be willing to modify the rule so that it can be configured to require one strategy or the later, warns on an incorrect strategy, and autofixes to the correct one.

It would be fine, however, to default to "both strategies are OK" via an array of both options, in either order, and to just pick the first one for autofix, and then the user could swap the order if they wanted the other autofix but both to be acceptable.

Reply:
Sounds like a good approach. I'll try to address the proposed changes during the week.

Reply:
This should be addressed now!